src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs

```diff
@@ -12,14 +12,434 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics
+                //         _f = ref (new int[1])[0];
+                Diagnostic(ErrorCode.ERR_RefLocalOrParamExpected, "_f").WithLocation(7, 9));
+        }
+
+        [Fact]
+        public void RefReassignOperatorResult()
+        {
+            var comp = CreateStandardCompilation(@"
+class C
+{
+    void M()
+    {
+        int x = 0;
+        (2 + 3) = ref x;
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (7,10): error CS8355: The left-hand side of a ref assignment must be a ref local or parameter.
+                //         (2 + 3) = ref x;
+                Diagnostic(ErrorCode.ERR_RefLocalOrParamExpected, "2 + 3").WithLocation(7, 10));
+        }
+
+        [Fact]
+        public void RefReassignToValue()
+        {
+            var comp = CreateStandardCompilation(@"
+class C
+{
+    void M()
+    {
+        int x = 0;
+        int y = 0;
+        x = ref y;
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (8,9): error CS8355: The left-hand side of a ref assignment must be a ref local or parameter.
+                //         x = ref y;
+                Diagnostic(ErrorCode.ERR_RefLocalOrParamExpected, "x").WithArguments("x").WithLocation(8, 9));
+        }
+
+        [Fact]
+        public void RefReassignWithReadonly()
+        {
+            var comp = CreateStandardCompilation(@"
+class C
+{
+    void M()
+    {
+        int x = 0;
+        ref int rx = ref x;
+        ref readonly int rrx = ref x;
+        rx = ref rrx;
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (9,18): error CS1510: A ref or out value must be an assignable variable
+                //         rx = ref rrx;
+                Diagnostic(ErrorCode.ERR_RefLvalueExpected, "rrx").WithLocation(9, 18));
+        }
+
+        [Fact]
+        public void RefReassignToRefReturn()
+        {
+            var comp = CreateStandardCompilation(@"
+class C
+{
+    void M()
+    {
+        int x = 0;
+        ref int rx = ref M2();
+        M2() = ref x;
+    }
+    ref int M2() => ref (new int[1])[0];
+}");
+            comp.VerifyDiagnostics(
+                // (8,9): error CS8355: The left-hand side of a ref assignment must be a ref local or parameter.
+                //         M2() = ref x;
+                Diagnostic(ErrorCode.ERR_RefLocalOrParamExpected, "M2()").WithLocation(8, 9));
+        }
+
+        [Fact]
+        public void RefReassignToProperty()
+        {
+            var comp = CreateStandardCompilation(@"
+class C
+{
+    void M()
+    {
+        int x = 0;
+        ref int rx = ref P;
+        P = ref x;
+    }
+    ref int P
+    {
+        get => ref (new int[1])[0];
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (8,9): error CS8355: The left-hand side of a ref assignment must be a ref local or parameter.
+                //         P = ref x;
+                Diagnostic(ErrorCode.ERR_RefLocalOrParamExpected, "P").WithLocation(8, 9));
         }
```

Comment:
Can you also do some testing (manual is probably sufficient) on IDE? Formatting of `x=ref y;`, colorization, completion, extract method, extract local, inlining temp, ... File follow-up issues as needed.
Also, `(x, y) = ref (z, w); // some error(s)`.

Reply:
Consider testing `x = out y;` and `x = in z;` (both parse errors)

---
In reply to: [143597161](https://github.com/dotnet/roslyn/pull/22608#discussion_r143597161) [](ancestors = 143597161)